File: Algorithms/binaryTreeInorderTraversal.cpp
class Solution {
public:
    vector<int> inorderTraversal(TreeNode *root) {
        vector<int> v;
        stack<TreeNode*> stack;
        TreeNode *p = root;
        while (!stack.empty() || p != NULL) {
            if (p != NULL) {
                stack.push(p);
                p = p->left;
            }
            else {
                p = stack.top();
                stack.pop();
                v.push_back(p->val);
                p = p->right;
            }
        }
        return v;
    }
    /*
    // recursive
    vector<int> inorderTraversal(TreeNode *root) {
        vector<int> v;
        if (root == NULL) {
            return v;
        }
        vector<int> left = inorderTraversal_recursive(root->left);
        for (int i=0; i < left.size(); i++) {
            v.push_back(left[i]);
        }
        v.push_back(root->val);
        vector<int> right = inorderTraversal_recursive(root->right);
        for (int i=0; i < right.size(); i++) {
            v.push_back(right[i]);
        }
        return v;
    }
    */
};

File: Algorithms/binaryTreeInorderTraversal.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> v = new ArrayList<Integer>();
        Stack<TreeNode> stack = new Stack<TreeNode>();
        TreeNode p = root;
        while (stack.size() != 0 || p != null) {
            if (p != null) {
                stack.add(p);
                p = p.left;
            }
            else {
                p = stack.peek();
                stack.pop();
                v.add(p.val);
                p = p.right;
            }
        }
        return v;
    }
    public List<Integer> inorderTraversal_recursive(TreeNode root) {
        List<Integer> v = new ArrayList<Integer>();
        if (root == null) {
            return v;
        }
        ArrayList<Integer> left = inorderTraversal_recursive(root.left);
        for (int i=0; i < left.size(); i++) {
            v.add(left.get(i));
        }
        v.add(root.val);
        ArrayList<Integer> right = inorderTraversal_recursive(root.right);
        for (int i=0; i < right.size(); i++) {
            v.add(right.get(i));
        }
        return v;
    }
}
